### ISSUE TITLE
Stack traces NEVER available for any errors, for no reason

### ISSUE THREAD

**Issue by php4fan** [newrelic/newrelic-browser-agent#1004]
https://forum.newrelic.com/s/hubtopic/aAXPh0000001XJhOAM/missing-stack-traces-in-js-errors

Read the thread and you'll see that something MUST be wrong in NewRelic.

Sentry captures the stack traces out-of-the-box just fine, for the same errors on the same app. So if the NewRelic browser agent can't do the same, either it's broken, or if it requires any extra configuration, then it's poorly designed.

I'm sorry but I've already spent too much time collecting and providing information, so just take it from there.
I can answer questions if you need any additional information that is not already in the linked forum thread.

**Comment by metal-messiah** [comment]
Thanks for reporting this.  We are working through investigating this now.  I’ll keep this issue updated with findings as they come.

-- So i went to the error link mentioned here

> "Please start from this error:
>     <omitted> (it's the same I mentioned in a previous comment).
>     This one certainly doesn't have a valid reason for not having a stack trace (and it did produce stack traces in Sentry).
>     It can be triggered on purpose by visiting: 
> 
>      
>     This application still is in "Pro" mode. The other one I've switched back to "Lite" because it was generating too much usage.  I only temporarily tried Sentry to see if it could capture error traces for the same application (and it did, at least for some errors)."

But that error generates a stack trace for me just fine. 

![image](https://github.com/newrelic/newrelic-browser-agent/assets/24296585/df07f9eb-f042-4310-8bbe-60245d6c86e8)
![image](https://github.com/newrelic/newrelic-browser-agent/assets/24296585/3e78f211-6ab9-41a3-87fb-55127391c52f)

Jordan Porter 
12 minutes ago

Thanks for reporting this.  We are working through investigating this now.  I’ll keep this issue updated with findings as they come.

-- So i went to the error link mentioned here

    "Please start from this error:
    <omitted> (it's the same I mentioned in a previous comment).
    This one certainly doesn't have a valid reason for not having a stack trace (and it did produce stack traces in Sentry).
    It can be triggered on purpose by visiting: 

     
    This application still is in "Pro" mode. The other one I've switched back to "Lite" because it was generating too much usage.  I only temporarily tried Sentry to see if it could capture error traces for the same application (and it did, at least for some errors)."

But that error generates a stack trace for me just fine. 

image-20240430-170708.png

 
image-20240430-170713.png

still looking into why some of the errors have a stackTrace = ''

**Comment by metal-messiah** [comment]
Some other findings (mainly confirming the findings in the forum thread) -- The following are third party scripts being loaded on https://clubxmax.com/perdida-de-aceite-por-la-bomba-de-agua-t20154.html --> (one of the sites shown to throw errors with no stack in the [query history](https://onenr.io/0gR7OxngxRo))
- https://unpkg.com/leaflet@1.7.1/dist/leaflet.js
    - access-control-allow-origin[: ](https://unpkg.com/leaflet@1.7.1/dist/leaflet.js)'*'
- https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/jquery-ui.min.js
    - access-control-allow-origin: '*'
- https://www.gstatic.com/charts/loader.js (which spawns a bunch more child asset calls to this domain)
    - no access-control-allow-origin response header found on any of these
- https://www.googletagmanager.com/gtag/js?id=G-1R1638E7CF
    - access-control-allow-origin: '*'

**Comment by metal-messiah** [comment]
shot in the dark to investigate, but i wonder if some browsers have some flavor of a "high security" mode that prevents anonymous CORS calls from exposing any stack trace info

**Comment by php4fan** [comment]
@metal-messiah 
> But that error generates a stack trace for me just fine.

Well that's astonishing. This is what I see:

![image](https://github.com/newrelic/newrelic-browser-agent/assets/13969584/50920edd-2d87-43e9-92e0-50a6d01b5032)

Then, correct me if I'm wrong, it means NewRelic is capturing the stack traces just fine but apparently there's some stupid issue in the UI that doesn't show the stack trace to me??

> but i wonder if some browsers have some flavor of a "high security" mode that prevents anonymous CORS calls from exposing any stack trace info

It's not specific to "some browsers". Even when I myself trigger the most trivial error from Google Chrome, the stack trace is not available to me in NewRelic. I have never been able to see a single JavaScript stack trace in NewRelic.

**Comment by metal-messiah** [comment]
> Then, correct me if I'm wrong, it means NewRelic is capturing the stack traces just fine but apparently there's some stupid issue in the UI that doesn't show the stack trace to me??
To rule that out, are you able to see a stack trace from the event I observed on my end?
https://onenr.io/0Bj3xzmAvQX

**Comment by metal-messiah** [comment]
If we cant get a resolution here in GitHub, I will get with our support team and get a more formal support ticket in place so we can make sure to find a resolution and have the ability to track success

**Comment by metal-messiah** [comment]
When you have time to look into this, would you see if you would be able to send me a HAR file of a page where you throw an error?  I'd like to rule out if the browser agent itself is the culprit for not sending the stack trace info, or if the backend is sanitizing that value in some way.

**Comment by php4fan** [comment]
> To rule that out, are you able to see a stack trace from the event I observed on my end?

Oh that's funny, yes I do see the stack trace there. It's the first time I get to see a browser stack trace in NewRelic.

I see now that if I trigger the error in Firefox (as opposed to Chrome) I get the stack trace too.

So they are not always missing, just most of the time.

> About 2/3rds of the errors that are reported have a valid stack trace

So, not even close to most of the time. It's surprising how unlucky I've been, NEVER seeing a single stack trace up to now (and I assure you I opened a lot of errors).

Can you try to trigger the error from Chrome?

> There are definitely some errors coming through with no stack trace from the agent

Note that there might be a subset of legitimate cases that meet some of the documented causes. But there definitely are errors that should have a stack trace, like the one above, and I suspect a few others.

> would you see if you would be able to send me a HAR file

[clubxmax.com.har.zip](https://github.com/newrelic/newrelic-browser-agent/files/15171038/clubxmax.com.har.zip)

**Comment by metal-messiah** [comment]
Thanks for sharing that, i'll be working through digging into the HAR findings (which have no stack trace at the agent level) as well as looking at Chrome v. Firefox.  Keep you posted here

**Comment by metal-messiah** [comment]
I have replicated that Chrome itself does not add an explicit stack to the error that is emitted.  That said, there are some things the browser agent could do to decorate the error we report with at least some basic information, such as the line and column number of the error, which at the moment is being skipped for these particular cases where the stack is missing. 

I have added a bug ticket to our log to address the "skipping" behavior when a stack trace is not found.  I will continue looking for reasons `why` chrome is not attaching the stack trace, when Firefox does successfully.

**Comment by metal-messiah** [comment]
#1014

**Comment by metal-messiah** [comment]
Decorating errors that have no stack trace with basic line and column info should be mitigated to the best of our abilities in release https://github.com/newrelic/newrelic-browser-agent/releases/tag/v1.258.1

**Comment by php4fan** [comment]
This still does not fix the issue that errors don't have stack traces in Chrome, though.

It is not a Chrome issue, it's a NewRelic issue that for some reason is specific to Chrome. Sentry is able to report full stack traces for the same errors, from Chrome, so it's not inherently impossible.

**Comment by metal-messiah** [comment]
@php4fan would you mind sharing what the stack trace looks like in Sentry? I have personally validated that with the error you shared (the intentional `lsdflsnfv is not defined`), the `error event` that Chrome emits does not have a stack trace at all, likely tied some third party cross origin handling. The `error event` can be populated differently than what chrome shows in the console. When the `error event's` error object has a stack, it reports as expected, as seen in the Firefox example linked above (which does supply a valid stack for the error). When the `error event's` error object is missing a stack, the only thing we would be able to extract from that event would be the message, col, line, and file name if present. Would be curious to see how Sentry differs there to help point us to something we are missing.

**Comment by php4fan** [comment]
> would you mind sharing what the stack trace looks like in Sentry?

Sure, here it is:

![image](https://github.com/newrelic/newrelic-browser-agent/assets/13969584/2544a0d5-5d6f-4198-9aba-3770c4b4e15e)

**Comment by php4fan** [comment]
OMG I found out I had this in my code: 

```
Error.stackTraceLimit = undefined;
```
which was intended to set the limit to infinite. At the time I put it there, several years ago, it definitely worked, but I guess at some point they changed the behavior. Now setting it to undefined apparently has the opposite effect, it disable capturing stack traces entirely (it's not documented actually, but it's the documented effect of setting it to zero), at least as far as `<errorEvent>.error.stack` is concerned.

I think Sentry explicitly sets `Error.stackTraceLimit = 50` at some point, and I guess in my tests with sentry, theirs code was being executed after than mine, overriding it.

I'm not sure if NewRelic also sets `stackTraceLimit` but I was setting it after the NewRelic browser agent script had run.

I've removed it and now I'm getting Stack traces in NewRelic.
